File: java/src/main/MaximumSubarraySumWithLengthDivisibleByK.java
import util.ArrayUtils;

/**
 * 3381. 长度可被 K 整除的子数组的最大元素和：https://leetcode.cn/problems/maximum-subarray-sum-with-length-divisible-by-k
 */
public class MaximumSubarraySumWithLengthDivisibleByK {

    public long maxSubarraySum(int[] nums, int k) {
        long[] sumArr = new long[nums.length];
        long max = Long.MIN_VALUE;
        long sum = 0L;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (i + 1 < k) {
                sumArr[i] = Long.MIN_VALUE;
                continue;
            }
            long x = sum;
            int prevIndex = i - k;
            if (prevIndex >= 0 && sumArr[prevIndex] > 0) {
                x += sumArr[prevIndex];
            }
            sumArr[i] = x;
            max = Math.max(max, x);
            sum -= nums[i - k + 1];
        }
        return max;
    }

    public static void main(String[] args) {
        MaximumSubarraySumWithLengthDivisibleByK app = new MaximumSubarraySumWithLengthDivisibleByK();
        System.out.println(app.maxSubarraySum(ArrayUtils.array("[1,2]"), 1)); // 3
        System.out.println(app.maxSubarraySum(ArrayUtils.array("[-5,1,2,-3,4]"), 2)); // 4
    }

}

File: c/maximum-subarray-sum-with-length-divisible-by-k.c
/**
 * 3381. 长度可被 K 整除的子数组的最大元素和：https://leetcode.cn/problems/maximum-subarray-sum-with-length-divisible-by-k
 */

#include <limits.h>
#include <string.h>

#define MAX(a,b) (((a)>(b))?(a):(b))

long long dfs(int* nums, int i, int k, int n, long long* memo, long long* preSum) {
    if (i < 0 || i + 1 < k) {
        return LLONG_MIN / 2;
    }
    if (memo[i] != LLONG_MIN) {
        return memo[i];
    }
    long long sum = preSum[i + 1] - preSum[i - k + 1];
    return memo[i] = sum + MAX(0, dfs(nums, i - k, k, n, memo, preSum));
}

long long maxSubarraySum(int* nums, int n, int k) {
    long long memo[n];
    long long preSum[n + 1];
    preSum[0] = 0;
    for (int i = 0; i < n; i++) {
        memo[i] = LLONG_MIN;
        preSum[i + 1] = preSum[i] + nums[i];
    }
    long long ans = LLONG_MIN;
    for (int i = 0; i < n; i++) {
        ans = MAX(ans, dfs(nums, i, k, n, memo, preSum));
    }
    return ans;
}

long long maxSubarraySum2(int* nums, int n, int k) {
    long long preSum[n + 1];
    preSum[0] = 0;
    for (int i = 0; i < n; i++) {
        preSum[i + 1] = preSum[i] + nums[i];
    }
    long long f[n + 1];
    for (int i = 0; i < k; ++i) {
        f[i] = LLONG_MIN / 2;
    }

    long long ans = LLONG_MIN;
    for (int i = k - 1; i < n; i++) {
        f[i + 1] = preSum[i + 1] - preSum[i - k + 1] + MAX(0, f[i - k + 1]);
        ans = MAX(ans, f[i + 1]);
    }
    return ans;
}

long long maxSubarraySum3(int* nums, int n, int k) {
    long long preSum[n + 1];
    preSum[0] = 0;
    long long f[n + 1];
    for (int i = 0; i < k; ++i) {
        f[i] = LLONG_MIN / 2;
        preSum[i + 1] = preSum[i] + nums[i];
    }

    long long ans = LLONG_MIN;
    for (int i = k - 1; i < n; i++) {
        preSum[i + 1] = preSum[i] + nums[i];
        f[i + 1] = preSum[i + 1] - preSum[i - k + 1] + MAX(0, f[i - k + 1]);
        ans = MAX(ans, f[i + 1]);
    }
    return ans;
}

long long maxSubarraySum4(int* nums, int n, int k) {
    long long f[k];
    long long preSum[k + 1];
    preSum[0] = 0;
    for (int i = 0; i < k; ++i) {
        f[i] = LLONG_MIN / 2;
        preSum[i + 1] = preSum[i] + nums[i];
    }

    long long ans = LLONG_MIN;
    for (int i = k - 1; i < n; i++) {
        preSum[(i + 1) % (k + 1)] = preSum[i % (k + 1)] + nums[i];
        f[(i + 1) % k] = preSum[(i + 1) % (k + 1)] - preSum[(i - k + 1) % (k + 1)] + MAX(0, f[(i - k + 1) % k]);
        ans = MAX(ans, f[(i + 1) % k]);
    }
    return ans;
}